tests/test_siemens_td3511.py

```diff
@@ -0,0 +1,139 @@
+    data_str.append(b'1.8.2(18696.758*kWh)\r\n')
+    data_str.append(b'2.8.0(22309.592*kWh)\r\n')
+    data_str.append(b'2.8.1(16717.051*kWh)\r\n')
+    data_str.append(b'2.8.2(5592.541*kWh)\r\n')
+    data_str.append(b'3.8.1(68.340*kvarh)\r\n')
+    data_str.append(b'4.8.1(29332.587*kvarh)\r\n')
+    data_str.append(b'0.9.1(21:10:29)\r\n')
+    data_str.append(b'0.9.2(24-03-21)\r\n')
+    data_str.append(b'1.7.0(0.386*kW)\r\n')
+    data_str.append(b'2.7.0(0.000*kW)\r\n')
+    data_str.append(b'3.7.0(0.000*kvar)\r\n')
+    data_str.append(b'4.7.0(0.727*kvar)\r\n')
+    data_str.append(b'0.9.1(21:10:29)\r\n')
+    data_str.append(b'0.9.2(24-03-21)\r\n')
+    data_str.append(b'14.7(49.96*Hz)\r\n')
+    data_str.append(b'32.7(238.3*V)\r\n')
+    data_str.append(b'52.7(240.2*V)\r\n')
+    data_str.append(b'72.7(240.0*V)\r\n')
+    data_str.append(b'31.7(1.58*A)\r\n')
+    data_str.append(b'51.7(1.50*A)\r\n')
+    data_str.append(b'71.7(0.77*A)\r\n')
+    data_str.append(b'81.7.4(-80.7*Deg)\r\n')
+    data_str.append(b'81.7.15(-33.3*Deg)\r\n')
+    data_str.append(b'81.7.26(-74.5*Deg)\r\n')
+    data_str.append(b'!\r\n')
+    return data_str
+
+
+@pytest.mark.skipif(sys.version_info < (3, 8), reason="Python3.7 does not support AsyncMock.")
+@pytest.mark.asyncio
+async def test_siemens_td3511_parse_and_provide_unencrypted_data(mocker: MockerFixture,
+                                                         unencrypted_valid_data_siemens: List[bytes]):
+    observer = mocker.stub("collector_mock")
+    observer.mock_add_spec(['notify'])
+    serial_mock = mocker.patch("smartmeter_datacollector.smartmeter.siemens_td3511.SiemensSerialReader",
+                               autospec=True).return_value
+    serial_mock.TERMINATION_FLAG = b'!\r\n'
+    serial_mock.timestamp=datetime.now()
+    meter = SiemensTD3511("/test/port")
+    meter.register(observer)
+
+    def data_received():
+        for frame in unencrypted_valid_data_siemens:
+            meter._data_received(frame)
+    serial_mock.start_and_listen.side_effect = data_received
+
+    await meter.start()
+
+    serial_mock.start_and_listen.assert_awaited_once()
+    observer.notify.assert_called_once()
+    values = observer.notify.call_args.args[0]
+    assert isinstance(values, list)
+    assert any(data.type == MeterDataPointTypes.ACTIVE_POWER_P.value for data in values)
+    assert any(data.type == MeterDataPointTypes.ACTIVE_POWER_N.value for data in values)
+    assert any(data.type == MeterDataPointTypes.REACTIVE_POWER_P.value for data in values)
+    assert any(data.type == MeterDataPointTypes.REACTIVE_POWER_N.value for data in values)
+    assert any(data.type == MeterDataPointTypes.ACTIVE_ENERGY_P.value for data in values)
+    assert any(data.type == MeterDataPointTypes.ACTIVE_ENERGY_N.value for data in values)
+    assert all(data.source == "110002267" for data in values)
+    assert all(data.timestamp.strftime(r"%m/%d/%y %H:%M:%S") == "03/21/24 20:10:29" for data in values)
```

Comment:
You might have an issue with timezones. That is why the CI pipeline does not pass.
What timezone do the timestamps coming from the meter have? Local time or UTC?

With other meters we discovered that usually local time is sent by the meter using the CII port (with MBus, HDLC, DLMS)...

Reply:
I have to convert back to the local system timezone. This should make it work on any system timezone.